src/sqlfluff/core/rules/surrogates/segments.py

```diff
@@ -0,0 +1,130 @@
+"""Surrogate class for working with Segment collections."""
+from typing import Any, Callable, List, Optional, Sequence, Type, TypeVar
+
+from sqlfluff.core.parser import BaseSegment
+from sqlfluff.core.templaters.base import TemplatedFile
+from sqlfluff.core.rules.surrogates.raw_file_slice import RawFileSlices
+
+Predicate = TypeVar("Predicate", str, Type, Callable[[BaseSegment], bool])
+
+
+class Segments(tuple):
+    """Encapsulates a sequence of one or more BaseSegments.
+
+    The segments may or may not be contiguous in a parse tree.
+    Provides useful operations on a sequence of segments to simplify rule creation.
+    """
+
+    def __new__(cls, templated_file, *segments):
+        """Override new operator."""
+        return super(Segments, cls).__new__(cls, segments)
+
+    def __init__(self, templated_file: Optional[TemplatedFile], *_: BaseSegment):
+        self.templated_file = templated_file
+
+    def all(self, *predicates: Predicate) -> bool:
+        """Do all the segments match?"""
+        cp = _CompositePredicate(*predicates)
+        return all(cp(s) for s in self)
+
+    def any(self, *predicates: Predicate) -> bool:  # pragma: no cover
+        """Do any of the segments match?"""
+        cp = _CompositePredicate(*predicates)
+        return any(cp(s) for s in self)
+
+    @property
+    def raw_slices(self) -> RawFileSlices:
+        """Raw slices of the segments."""
+        if not self.templated_file:  # pragma: no cover
+            raise ValueError(
+                'Segments.raw_slices: "templated_file" property is required.'
+            )
+        raw_slices = set()
+        for s in self:
+            source_slice = s.pos_marker.source_slice
+            raw_slices.update(
+                self.templated_file.raw_slices_spanning_source_slice(source_slice)
+            )
+        return RawFileSlices(self.templated_file, *raw_slices)
+
+    def children(self, *predicates: Predicate) -> "Segments":
+        """Returns an object with children of the segments in this object."""
+        child_segments: List[BaseSegment] = []
+        cp = _CompositePredicate(*predicates)
+        for s in self:
+            for child in s.segments:
+                if cp(child):
+                    child_segments.append(child)
+        return Segments(self.templated_file, *child_segments)
+
+    def first(self, *predicates) -> Optional[BaseSegment]:
+        """Returns the first segment (if any) that satisfies the predicates."""
+        cp = _CompositePredicate(*predicates)
+        for s in self:
+            if cp(s):
+                return s
+        return None
+
+    def last(self, *predicates) -> Optional[BaseSegment]:
+        """Returns the last segment (if any) that satisfies the predicates."""
+        cp = _CompositePredicate(*predicates)
+        for s in reversed(self):
+            if cp(s):
+                return s
+        return None
+
+    def apply(self, fn: Callable[[BaseSegment], Any]) -> List[Any]:
```

Comment:
Do we maybe want a `reversed` method?

```python
def reversed(self) -> "Segments":
    return reversed(self)
```
```python
raw_stack = context.surrogates.raw_stack
reversed_raw_stack = raw_stack.reversed()
```
^ Feel like that's a fairly common thing someone may do.
(N.B. tuples are immutable so this returns a new instance of Segments i.e. `id(reversed_raw_stack) != id(raw_stack)`, which I like)

Reply:
Another thing that would be worth checking:
Since we inherit from tuple I assume we also inherit the __add__ method?
So can we do something like `segments_a + segments_b`? 
Could be nice

Reply:
following on from above, I know I've done the thing before where I've made a "complete_stack" which could be something like the raw_stack + plus the current segment. Could we also have an append method for something like:
```python
raw_stack = context.surrogates.raw_stack
complete_stack = raw_stack.append(context.segment)
```
I know tuple doesn't have an append method itself, but you can effectively mock it with __add__
```python
def append(self, segment):
    return self + self.__class__(segment)
```

Reply:
I'll add the `reversed()` method now. I'll have to check on `add()`. We probably inherit it, but I'm not 100% sure it'll work correctly. From my experience with operator overloading in other languages, it seems likely that either `my_tuple + segments` or `segments + my_tuple` will return a bare `tuple`, not a `Segments` object.

`append()` will take some thought. With lists, it modifies the list **in place**. As written above, it looks like it would return a new `Segments` object. I don't think we'll be able to modify the base tuple in place, because tuples are immutable. It may be possible to inherit from `list` instead.

I'll add these to the "to do" list.

Reply:
yeah I know tuples are immutable but we don't need to modify in place.

assigning to a new Segments instance is fine:
`complete_stack = raw_stack.append(context.segment)`

Reply:
I agree that'll work. I think it would be a confusing API because the name matches the similar function on `list`, which does modify in place. Imagine the poor rule developer: they assume it modified the object in place, so they didn't assign the result to a variable. The result is simply discarded, and they are left with the unmodified object. 🤯

I think the right solution is to inherit from `list` instead. I'll make that change when I have a chance. 👍